src/storage/mvcc/reader.rs

```diff
@@ -401,9 +406,158 @@ impl MvccReader {
                     },
                 }
             };
-            if let Some(v) = self.get(&key, ts)? {
+
+            if let Some(v) = self.reverse_get_impl(&key, ts)? {
                 return Ok(Some((key, v)));
             }
+
+            // reverse_get_impl may call write_cursor's prev.
+            if !self.write_cursor.as_mut().unwrap().valid() {
+                write_valid = false;
+            }
+        }
+    }
+
+    fn reverse_get_impl(&mut self, user_key: &Key, ts: u64) -> Result<Option<Value>> {
+        assert!(self.lock_cursor.is_some());
+        assert!(self.write_cursor.is_some());
+
+        // Check lock.
+        match self.isolation_level {
+            IsolationLevel::SI => {
+                let lock = {
+                    let l_cur = self.lock_cursor.as_mut().unwrap();
+                    if l_cur.valid() && l_cur.key() == user_key.encoded().as_slice() {
+                        self.statistics.lock.processed += 1;
+                        Some(Lock::parse(l_cur.value())?)
+                    } else {
+                        None
+                    }
+                };
+                if let Some(l) = lock {
+                    self.check_lock_impl(user_key, ts, l)?;
+                }
+            }
+            IsolationLevel::RC => {}
+        }
+
+        // Get value for this user key.
+        let mut round = 0;
+        let mut lastest_version = (None /* start_ts */, None /* short value */);
+        let mut last_handled_key: Option<Vec<u8>> = None;
+        loop {
+            if !self.write_cursor.as_mut().unwrap().valid() {
+                return self.get_value(user_key, lastest_version.0, lastest_version.1);
+            }
+
+            // When there are many versions for this user key, after several tries,
+            // we will use seek to locate the right position. But this will turn around
+            // the write cf's iterator's direction inside RocksDB, and the next user key
+            // need to turn back the direction to backward. As we have tested, turn around
+            // iterator's direction from forward to backward is as expensive as seek in
+            // RocksDB, so don't set REVERSE_SEEK_BOUND too small.
+            if round >= REVERSE_SEEK_BOUND {
```

Comment:
It seems to use `for _ in 0.. REVERSE_SEEK_BOUND` instead of `loop` will be better in performance? @BusyJay